profile-widget/profilewidget2.cpp

```diff
@@ -1440,12 +1449,25 @@ void ProfileWidget2::contextMenuEvent(QContextMenuEvent *event)
 	QPointF scenePos = mapToScene(mapFromGlobal(event->globalPos()));
 	qreal sec_val = timeAxis->valueAt(scenePos);
 	int seconds = (sec_val < 0.0) ? 0 : (int)sec_val;
-	if (current_dive && current_dive->cylinders.nr > 1) {
+	DiveEventItem *item = dynamic_cast<DiveEventItem *>(sceneItem);
+
+	// Add or edit Gas Change
+	if (item && event_is_gaschange(item->getEvent())) {
+		int eventTime = item->getEvent()->time.seconds;
+		QMenu *gasChange = m.addMenu(tr("Edit Gas Change"));
+		for (int i = 0; i < current_dive->cylinders.nr; i++) {
+			cylinder_t *cylinder = &current_dive->cylinders.cylinders[i];
+			QString label = printCylinderDescription(i, cylinder);
+			gasChange->addAction(label, [this, i, eventTime] { changeGas(i, eventTime); });
+		}
+	} else if (current_dive && current_dive->cylinders.nr > 1) {
 		// if we have more than one gas, offer to switch to another one
 		QMenu *gasChange = m.addMenu(tr("Add gas change"));
-		for (int i = 0; i < current_dive->cylinders.nr; i++)
-			gasChange->addAction(QString(current_dive->cylinders.cylinders[i].type.description) + tr(" (cyl. %1)").arg(i + 1),
-					     [this, i, seconds] { changeGas(i, seconds); });
+		for (int i = 0; i < current_dive->cylinders.nr; i++) {
+			cylinder_t *cylinder = &current_dive->cylinders.cylinders[i];
```

Comment:
Since you're changing this, perhaps switch to the more palatable `get_cylinder(current_dive, i)`.

Reply:
Seems a bit inefficient to have a loop which then uses an accessor function which re-iterates over the same array, just to ensure a check is made which doesn't apply in this situation,  I know this isn't performance-critical code, but still..

All that get_cylinder() currently does is a check in the Planner due to some hack (according to the comments) - however, the Planner doesn't even have a right-click menu, so this code will never even get triggered in the Planner.

Happy to change it regardless if you really think it improves the codebase. Although the "correct" way to fix it would be to properly encapsulate any types whose data should not be accessed directly. That would then force all code accessing the data to be updated instead of having an inconsistent codebase where some parts use accessors and other parts direct access.

Reply:
> Seems a bit inefficient to have a loop which then uses an accessor function which re-iterates over the same array, just to ensure a check is made which doesn't apply in this situation, I know this isn't performance-critical code, but still..

This is an absolutely negligible cost compared to the string manipulation that you do in the next line - not even talking about the expense of showing the menu on the screen. That's a complete non-issue.

However, we get:
- More readable code.
- More flexible code - we can change the layout of the cylinders in the dive.
- Safer code - it does bound check. You say it "doesn't apply in this situation", which is correct, but only if you didn't make a typo (e.g. loop over "j" and access "i").
- More consistent code - the number of "get_cylinder()" calls outnumbers the direct accesses.

So overall this appears to be a win.

> All that get_cylinder() currently does is a check in the Planner due to some hack (according to the comments) - however, the Planner doesn't even have a right-click menu, so this code will never even get triggered in the Planner.

You misread the comment (probably because it was poorly written): The check is not there for the planner. The check was _relaxed_ for the planner: You can now access one past the last "official" number of cylinders. There is always one "reserve" cylinder, which represents air. It's a sad story.

> Although the "correct" way to fix it would be to properly encapsulate any types whose data should not be accessed directly. That would then force all code accessing the data to be updated instead of having an inconsistent codebase where some parts use accessors and other parts direct access.

I would only do this if we change the core to C++, because C++ has all the necessary encapsulation features. I'm against "simulating" C++ in C - it can be done, but is very annoying. We could then simply change to `std::vector<cylinder_t>`. I checked and there don't seem to be many direct accesses to the cylinders array left. I will write a cleanup commit.

Reply:
Sorry, yes I know the performance hit is absolutely negligible; it's just the _obviousness_ of traversing the array twice that triggered my OCD.

As for safer, definitely, no arguments.

As for more readable... well.. given we're iterating in a loop which does a deep access of the data structures, to me personally, it's less readable. More readable would be if a we also had a get_num_cylinders() or, ideally, a get_first_cylinder() and get_next_cylinder().  
Your clean-up commit primarily accesses which were iterating over the cylinders instead of just accessing a particular cylinder, so it seems this iteration over cylinders is quite a common pattern in the codebase.

Just my 2c.  Happy to make the change with no further argument needed :)

Reply:
Sure - feel free to add a `get_num_cylinders()` function or even a `for_each_cylinder` macro akin to the `for_each_dive` macro.

We *could* define `begin()` and `end()` functions for C++ code, so that one can write
```
   for(const cylinder_t &cyl: d->cylinders) {
   }
```

However, if we want to start introducing C++-isms into core, that would need a discussion first.